### ISSUE TITLE
Suggestion for an improvement to SPF Flattening

### ISSUE THREAD

**Issue by siacali** [StackExchange/dnscontrol#1961]
Right now the SPF flattening only deals with Flattening SPF TXT records.  Since the overall goal of flattening is to reduce lookups, would it make sense to "flatten" A records too (i.e. resolve them to an IP address)?

**Comment by tlimoncelli** [comment]
Ah!  That's an interesting idea.

While that would reduce the lookups, I think replacing hostnames with IP addresses is a risk I wouldn't impose on my employer's website.  I wouldn't recommend something that I wouldn't do myself.

**Comment by siacali** [comment]
How do you see that as much different than reducing includes to static lists (i.e. same effect)?

**Comment by tlimoncelli** [comment]
`A` records often change dynamically, or resolve to different IPs in different parts of the world.

**Comment by siacali** [comment]
I believe that geo-location based A records would break SPF if used in the SPF a: field regardless of flattening.  

Dynamic DNS strikes me as a corner-case for which the “workaround” is simply “don’t use flattening.”  

If I think about the union set of users who:

1. Use dynamic DNS for an SMTP server (probably home users or really small techie businesses) who need an explicit a:
2. Or are “surprised” because someone they allow to send on their behalf uses dynamic DNS through an include:
3. And have SPF records large enough to need flattening
4. And Use dnscontrol to manage their records
5. And fail to heed the inevitable advice that should be in the doc that says “don’t try to flatten dynamic DNS A records”

This doesn’t appear add up to a whole lot of risk, although I’d agree that it’s likely to be non-zero (because…people…).  Do you see a flaw in my reasoning?

BTW…I am debating only because, as a rabid dnscontrol fan who is in the process of using it to clean up and manage thousands of domains owned by a single company, this is something that would be super-helpful and cleaner in the main product rather than being implemented through one-off extension.

**Comment by tlimoncelli** [comment]
I agree with your logic.  The union of those people is very small and they have the option of turning off that particular feature.

I'm curious how many of your domains are hitting the 10-lookup limit and are in need of saving the extra lookups by using IPs instead of names.

I'll gladly accept a PR.

**Comment by tlimoncelli** [comment]
Cool cool.

By the way... one side-effect of using `SPF_BUILDER()` at Stackoverflow is that it let us put a comment by each SPF item. That comment included who requested or why the item existed. This make ownership much more visible. As a result, it was easier to clean up the configuration and remove a lot of obsolete items. I hope it works similarly for you too!

**Comment by tlimoncelli** [comment]
I've decided against flattening A records.  The SPF flattening code is stable and in maintenance mode. If someone wants to submit a PR, I'll consider it, but I'm not going to actively develop SPF any more.

**Comment by case-fastly** [comment]
> I'm also considering creating/submitting a back-end for markmonitor - that will take a bit longer.

I randomly found your comment here, and was curious if you've given this any more thought, or started tinkering with this idea?
